### ISSUE TITLE
Storing an array of objects at a top level EntityStore always returns a list of Objects with no array prototype

### ISSUE THREAD

**Issue by IAMtheIAM** [datorama/akita#448]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x] Bug report  
[ ] Performance issue
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request
[ ] Other... Please describe:
</code></pre>

## Current behavior
Storing an array of objects at a top level EntityStore always returns a list of Objects with no array prototype. Therefore, no array methods can be called on the returned query.

## Expected behavior

If data is stored as an array of objects, it should be returned as an array of objects with the array prototype methods.

## Minimal reproduction of the problem with instructions

Create any akita store. 
Use update to add an array of objects to the top level EntityStore.
Query the entity store and view the response. It inherits from the object prototype.

## What is the motivation / use case for changing the behavior?

Data in should be the same as data out. Currently, the only way to store an array of objects in an EntityStore and receive it back as an array of objects is to put it as a property on the top level entity, like this:

![image](https://user-images.githubusercontent.com/13709948/84681546-20418800-aee9-11ea-9de7-96a80a68b50a.png)

However, it seems more logical to put the array directly inside the UI Entity Store, `TempEvent` in this case, rather than inside `TempEvent.events`, like this:

![image](https://user-images.githubusercontent.com/13709948/84683994-d9559180-aeec-11ea-95c6-41bb1a375222.png)

## Environment

<pre><code>
Angular version: 9.x.x

Browser:
All browsers
 
For Tooling issues:
- Node version: 12.16.3
- Platform: Windows

**Comment by NetanelBasal** [comment]
Please provide code examples.

**Comment by IAMtheIAM** [comment]
Hello. What kind of code example would you like? I have explained the scenario and provided screenshots of the Entity Stores showing how it is expected to work, and how it actually works. I cannot provide the actual code since it is for a client's project.

Take any array of objects, `upsert` it into a UI Entity Store, then call `this.queryService.ui.selectEntity(entity)` on it. What you will be returned with is this:

![image](https://user-images.githubusercontent.com/13709948/84684806-0e161880-aeee-11ea-8927-78f27a5acc21.png)

Which is an object of objects, inheriting from the `Object` prototype, instead of the Array prototype. Therefore, it cannot be looped over in a `for` loop of any sort and throws an error.

This should be returning an array prototype.

**Comment by IAMtheIAM** [comment]
Allow me some time and I will create a basic repro.

**Comment by IAMtheIAM** [comment]
https://stackblitz.com/edit/akita-todos-app-et1tpg?file=src%2Fapp%2Ftodos%2Ftodos-page%2Ftodos-page.component.ts

This is the basic repro. I added a method inside `todos-page.component.ts` that runs on init.

However, your stackblitz base repro is broken. It constantly says 
```
Error in src/app/app.module.ts (7:33)
Cannot find module '@datorama/akita-ngdevtools'.
```

So I could not run the stackblitz, but the code example is in there. Please check.

**Comment by xbaun** [comment]
You are merging your data array with the entity itself. Provide a dedicated data field for your entity and the array will be preserved: 
```ts
export interface Todo {
   [key: string]: any;
   data: any[]; // <--
};
```
And then updating the field `data` of the entity:
```ts
this.todosStore.upsert('EntityArray', { data });
```

**Comment by IAMtheIAM** [comment]
Hi @xbaun - yes I am aware that if I put it on a property inside the entity, the array prototype is preserved, as I put in my first screenshot. 

But this issue was a question more about why that is required, meaning, what is the design reason as to why the entity itself cannot be an array of objects?

**Comment by xbaun** [comment]
TL;DR If you only want to store an array of arrays within an entity store, it's most likely that the state should be [normalized](https://redux.js.org/recipes/structuring-reducers/normalizing-state-shape). 

> But this issue was a question more about why that is required, meaning, what is the design reason as to why the entity itself cannot be an array of objects?

I would say it is more of a design decision based on the well-established redux pattern. A store is simply a tree of [substates](https://redux.js.org/introduction/three-principles#single-source-of-truth). Each substate itself is an object again. An entity store as such contains as root state an array field of entities, whose entries in turn represent the state of an entity (basically the root state of the entity): Thus, these states are objects that can be queried by assigning an entity id. The generic array type does not provide custom fields, so its type definition must be extended to use it as an entity state (at least an entity id should exist for consistency), which makes handling the entire state more complex and error-prone.

> In addition, I'm wondering if putting an array on a data property of a single Entity item is a good practice with Akita, or if it would be better to create a new Entity item for each of those objects of the array. What's your opinion on that?

It depends on how complex the data and their relationship to each other is. A rule of thumb is to [keep things simple](https://datorama.github.io/akita/docs/best-practices/#keep-it-simple). Nested states are fine, but at some point it's easier to use separate stores to query and manage complex data. In the end it comes down to keeping a state as flat as possible without overdoing it.

**Comment by IAMtheIAM** [comment]
Thanks for the explanation, I will take some time to understand that and use the info appropriately.
